guidelines/auth.md

```diff
@@ -0,0 +1,142 @@
+# Authentication and Authorization Guideline
+
+This guideline documents the current best practices for authentication and
+authorization in Kubeflow. Kubeflow Application Developers should use this guide
+to effectively integrate their applications with Kubeflow. This document answers
+the following questions:
+* Authentication: How do application know the identity of the user making a
+  request?
+* Authorization: How do applications act on behalf of the user?
+
+
+## Introduction
+
+Kubeflow leverages Kubernetes to build its core APIs (CRDs), starting out with a
+few operators. The multi-tenancy story at this level was straightforward: direct
+interaction with the API via kubectl or other client, with the K8s API Server
+taking care of authentication and authorization.
+
+Soon after, Kubeflow introduced web apps to complement those core APIs and offer
+a better UX. This pushes the authentication and authorization problem out of the
+Kubernetes API Server and into the web apps.
+
+
+## Kubeflow Core (CRDs)
+Kubeflow started by introducing CRDs, providing an API on-top of Kubernetes.
+Interaction directly with the K8s API-Server via kubectl and
+authentication/authorization is a solved issue.
+
+
+### Authentication
+
+Whatever the [Kubernetes API-Server
+supports](https://kubernetes.io/docs/reference/access-authn-authz/authentication/):
+* ServiceAccount token
+* OIDC token
+* X509 Cert
+* Webhook
+* ...
+
+
+### Authorization
+
+K8s RBAC: Can `IDENTITY` perform `ACTION` on `RESOURCE` in `NAMESPACE`?
+
+
+## Kubeflow Web Apps
+
+Kubeflow wants to build on top of the functionality provided by its controllers.
+To provide a better UX, Kubeflow started building Web Apps to ease the use of
+its controllers. However, the story for authentication and authorization
+changes:
+* Authentication is performed by a proxy, not the K8s API Server.
+* Authorization is performed by the Web App Backend, not the K8s API Server.
+
+However, we still want to make sure that the user has the same identity and the
+same RBAC permissions are consulted.
+
+![kubeflow request flow](images/kubeflow_request_flow.svg)
+
+
+### Authentication
+
+Authentication is done at a proxy (e.g., the Istio Gateway, Google IAP) in front
+of the web app and the result is propagated to the web app in a header, in a way
+similar to the [authenticating proxy](https://kubernetes.io/docs/reference/access-authn-authz/authentication/#authenticating-proxy)
+method. The authenticated entity can be a user (OIDC, Google IAP) or a
+ServiceAccount (TokenReview).
+
+At the moment only userid is added as a header. However, there are other
```

Comment:
Should we mention JWT's and cite https://jwt.io/ here? There's a couple points that might be worth calling out.
JWT's are bearer tokens so they need to be protected and they should be signed. In practice, this means we should be using ISTIO side cars to verify signatures and creating a zero trust network (using mTLS and encrypting traffic between all proxies). We don't do that everywhere.

Reply:
I would go with no mention of JWTs here, because I view it as an implementation detail of how authentication is done. Since this doc is about defining the contract to the application (find the user's identity in headers), how authentication is implemented (JWT, TLS certificate, opaque access token) is out-of-scope for this doc, because no matter what the implementation is, the application will only see the result in headers.
This would be an excellent discussion for a doc reviewing the infrastructure configuration (Istio, Gateways, External Proxies) of Kubeflow (and we should definitely start one of these).